In [230]:
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt   #Esto es para poder hacer gráficos
import seaborn as sns
import re

***FASE 1 y 2: UNIÓN Y LIMPIEZA DE DATOS***

In [83]:
df_13= pd.read_csv("datos/datos-2013.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_14= pd.read_csv("datos/datos-2014.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_15= pd.read_csv("datos/datos-2015.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_16= pd.read_csv("datos/datos-2016.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_17= pd.read_csv("datos/datos-2017.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_18= pd.read_csv("datos/datos-2018.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_19= pd.read_csv("datos/datos-2019.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_20= pd.read_csv("datos/datos-2020.csv", sep= ";")
pd.set_option("display.max_columns", None)
df_21= pd.read_csv("datos/datos-2021.csv", sep= ";")
pd.set_option("display.max_columns", None)
#Falta abrir el diccionario de datos


In [84]:
df_dicc= pd.read_csv("datos/diccionario_datos.csv", sep= ";;", encoding= "latin-1")
pd.set_option("display.max_columns", None)

C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\3123453317.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df_dicc= pd.read_csv("datos/diccionario_datos.csv", sep= ";;", encoding= "latin-1")


In [85]:
df_dicc.head()

COLUNA                                         DESCRIÇÃO;
0                    NaN                                                  ;
1  Código Órgão Superior  Código do Órgão Superior responsável pela rece...
2                    NaN                                                  ;
3                    NaN  Nome do Órgão Superior responsável pela receita.;
4    Nome Órgão Superior                                                  ;

In [86]:
df_13.columns = [name.lower().strip() for name in df_13.columns]                        
df_14.columns = [name.lower().strip() for name in df_14.columns] 
df_15.columns = [name.lower().strip() for name in df_15.columns] 
df_16.columns = [name.lower().strip() for name in df_16.columns] 
df_17.columns = [name.lower().strip() for name in df_17.columns] 
df_18.columns = [name.lower().strip() for name in df_18.columns] 
df_19.columns = [name.lower().strip() for name in df_19.columns] 
df_20.columns = [name.lower().strip() for name in df_20.columns] 
df_21.columns = [name.lower().strip() for name in df_21.columns] #Cambios de nombre en todas las columnas de todos los df

***Comprobación de que tengan las mismas columnas***

In [87]:
df_13.columns.tolist()

['código órgão superior',
 'nome órgão superior',
 'código órgão',
 'nome órgão',
 'código unidade gestora',
 'nome unidade gestora',
 'categoria econômica',
 'origem receita',
 'espécie receita',
 'detalhamento',
 'valor previsto atualizado',
 'valor lançado',
 'valor realizado',
 'percentual realizado',
 'data lançamento',
 'ano exercício']

In [88]:
df_13.columns.tolist() == df_14.columns.tolist() ==df_15.columns.tolist() == df_16.columns.tolist() == df_17.columns.tolist() == df_18.columns.tolist() == df_19.columns.tolist() == df_20.columns.tolist() == df_21.columns.tolist() 

True

**COSAS PARA REPETIR EN TODAS**

In [89]:

df_13['nome órgão superior'] = df_13.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_13['nome órgão superior'].info() #Columna codigo y nombre


df_13["data lançamento"] = pd.to_datetime(df_13["data lançamento"]) 
df_13["data lançamento"]  #cambio a fecha

df_13["valor realizado"]= df_13["valor realizado"].str.replace(",",".")
df_13["valor realizado"] = list(map(float, df_13["valor realizado"]))
df_13["valor lançado"]= df_13["valor lançado"].str.replace(",",".")
df_13["valor lançado"] = list(map(float, df_13["valor lançado"]))
df_13["valor previsto atualizado"]= df_13["valor previsto atualizado"].str.replace(",",".")
df_13["valor previsto atualizado"] = list(map(float, df_13["valor previsto atualizado"]))
df_13["percentual realizado"]= df_13["percentual realizado"].str.replace(",",".")
df_13["percentual realizado"] = list(map(float, df_13["percentual realizado"]))    #cambio a float



<class 'pandas.core.series.Series'>
RangeIndex: 4498 entries, 0 to 4497
Series name: nome órgão superior
Non-Null Count  Dtype 
--------------  ----- 
4263 non-null   object
dtypes: object(1)
memory usage: 35.3+ KB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\910466304.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_13['nome órgão superior'] = df_13.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\910466304.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_13['nome órgão superior'] = df_13.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\910466304.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or spec

In [90]:
df_14.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4553 entries, 0 to 4552
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   código órgão superior      4331 non-null   float64
 1   nome órgão superior        2959 non-null   object 
 2   código órgão               4493 non-null   float64
 3   nome órgão                 4439 non-null   object 
 4   código unidade gestora     4332 non-null   float64
 5   nome unidade gestora       4392 non-null   object 
 6   categoria econômica        4502 non-null   object 
 7   origem receita             4487 non-null   object 
 8   espécie receita            4455 non-null   object 
 9   detalhamento               4506 non-null   object 
 10  valor previsto atualizado  4325 non-null   object 
 11  valor lançado              4375 non-null   object 
 12  valor realizado            4328 non-null   object 
 13  percentual realizado       4437 non-null   objec

In [91]:
df_14['nome órgão superior'] = df_14.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_14['nome órgão superior'].info() #Columna codigo y nombre


df_14["data lançamento"] = pd.to_datetime(df_14["data lançamento"]) 
df_14["data lançamento"]  #cambio a fecha

df_14["valor realizado"]= df_14["valor realizado"].str.replace(",",".")
df_14["valor realizado"] = list(map(float, df_14["valor realizado"]))
df_14["valor lançado"]= df_14["valor lançado"].str.replace(",",".")
df_14["valor lançado"] = list(map(float, df_14["valor lançado"]))
df_14["valor previsto atualizado"]= df_14["valor previsto atualizado"].str.replace(",",".")
df_14["valor previsto atualizado"] = list(map(float, df_14["valor previsto atualizado"]))
df_14["percentual realizado"]= df_14["percentual realizado"].str.replace(",",".")
df_14["percentual realizado"] = list(map(float, df_14["percentual realizado"]))    #cambio a float

<class 'pandas.core.series.Series'>
RangeIndex: 4553 entries, 0 to 4552
Series name: nome órgão superior
Non-Null Count  Dtype 
--------------  ----- 
4325 non-null   object
dtypes: object(1)
memory usage: 35.7+ KB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\2339529443.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_14['nome órgão superior'] = df_14.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\2339529443.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_14["data lançamento"] = pd.to_datetime(df_14["data lançamento"])


In [92]:
df_14.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4553 entries, 0 to 4552
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   código órgão superior      4331 non-null   float64       
 1   nome órgão superior        4325 non-null   object        
 2   código órgão               4493 non-null   float64       
 3   nome órgão                 4439 non-null   object        
 4   código unidade gestora     4332 non-null   float64       
 5   nome unidade gestora       4392 non-null   object        
 6   categoria econômica        4502 non-null   object        
 7   origem receita             4487 non-null   object        
 8   espécie receita            4455 non-null   object        
 9   detalhamento               4506 non-null   object        
 10  valor previsto atualizado  4325 non-null   float64       
 11  valor lançado              4375 non-null   float64       
 12  valor 

In [93]:
df_15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4523 entries, 0 to 4522
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   código órgão superior      4324 non-null   float64
 1   nome órgão superior        2940 non-null   object 
 2   código órgão               4351 non-null   float64
 3   nome órgão                 4361 non-null   object 
 4   código unidade gestora     4326 non-null   float64
 5   nome unidade gestora       4449 non-null   object 
 6   categoria econômica        4463 non-null   object 
 7   origem receita             4346 non-null   object 
 8   espécie receita            4320 non-null   object 
 9   detalhamento               4393 non-null   object 
 10  valor previsto atualizado  4297 non-null   object 
 11  valor lançado              4381 non-null   object 
 12  valor realizado            4351 non-null   object 
 13  percentual realizado       4322 non-null   objec

In [94]:
df_15['nome órgão superior'] = df_15.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_15['nome órgão superior'].info() #Columna codigo y nombre


df_15["data lançamento"] = pd.to_datetime(df_15["data lançamento"]) 
df_15["data lançamento"]  #cambio a fecha

df_15["valor realizado"]= df_15["valor realizado"].str.replace(",",".")
df_15["valor realizado"] = list(map(float, df_15["valor realizado"]))
df_15["valor lançado"]= df_15["valor lançado"].str.replace(",",".")
df_15["valor lançado"] = list(map(float, df_15["valor lançado"]))
df_15["valor previsto atualizado"]= df_15["valor previsto atualizado"].str.replace(",",".")
df_15["valor previsto atualizado"] = list(map(float, df_15["valor previsto atualizado"]))
df_15["percentual realizado"]= df_15["percentual realizado"].str.replace(",",".")
df_15["percentual realizado"] = list(map(float, df_15["percentual realizado"]))    #cambio a float

<class 'pandas.core.series.Series'>
RangeIndex: 4523 entries, 0 to 4522
Series name: nome órgão superior
Non-Null Count  Dtype 
--------------  ----- 
4308 non-null   object
dtypes: object(1)
memory usage: 35.5+ KB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\192372322.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_15['nome órgão superior'] = df_15.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\192372322.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_15["data lançamento"] = pd.to_datetime(df_15["data lançamento"])


In [95]:
df_15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4523 entries, 0 to 4522
Data columns (total 16 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   código órgão superior      4324 non-null   float64       
 1   nome órgão superior        4308 non-null   object        
 2   código órgão               4351 non-null   float64       
 3   nome órgão                 4361 non-null   object        
 4   código unidade gestora     4326 non-null   float64       
 5   nome unidade gestora       4449 non-null   object        
 6   categoria econômica        4463 non-null   object        
 7   origem receita             4346 non-null   object        
 8   espécie receita            4320 non-null   object        
 9   detalhamento               4393 non-null   object        
 10  valor previsto atualizado  4297 non-null   float64       
 11  valor lançado              4381 non-null   float64       
 12  valor 

In [96]:
df_16.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194533 entries, 0 to 194532
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   código órgão superior      190214 non-null  float64
 1   nome órgão superior        126446 non-null  object 
 2   código órgão               190506 non-null  float64
 3   nome órgão                 191144 non-null  object 
 4   código unidade gestora     185211 non-null  float64
 5   nome unidade gestora       191552 non-null  object 
 6   categoria econômica        192399 non-null  object 
 7   origem receita             186042 non-null  object 
 8   espécie receita            190129 non-null  object 
 9   detalhamento               186707 non-null  object 
 10  valor previsto atualizado  184806 non-null  object 
 11  valor lançado              188960 non-null  object 
 12  valor realizado            188766 non-null  object 
 13  percentual realizado       19

In [97]:
df_16['nome órgão superior'] = df_16.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_16['nome órgão superior'].info() #Columna codigo y nombre


df_16["data lançamento"] = pd.to_datetime(df_16["data lançamento"], format="%d/%m/%Y")   #He tenido que especificar que venía en otro formato distinto.
df_16["data lançamento"]  #cambio a fecha

df_16["valor realizado"]= df_16["valor realizado"].str.replace(",",".")
df_16["valor realizado"] = list(map(float, df_16["valor realizado"]))
df_16["valor lançado"]= df_16["valor lançado"].str.replace(",",".")
df_16["valor lançado"] = list(map(float, df_16["valor lançado"]))
df_16["valor previsto atualizado"]= df_16["valor previsto atualizado"].str.replace(",",".")
df_16["valor previsto atualizado"] = list(map(float, df_16["valor previsto atualizado"]))
df_16["percentual realizado"]= df_16["percentual realizado"].str.replace(",",".")
df_16["percentual realizado"] = list(map(float, df_16["percentual realizado"]))    #cambio a float

C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\3871943138.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_16['nome órgão superior'] = df_16.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))


<class 'pandas.core.series.Series'>
RangeIndex: 194533 entries, 0 to 194532
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
190203 non-null  object
dtypes: object(1)
memory usage: 1.5+ MB


In [98]:
df_16.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194533 entries, 0 to 194532
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      190214 non-null  float64       
 1   nome órgão superior        190203 non-null  object        
 2   código órgão               190506 non-null  float64       
 3   nome órgão                 191144 non-null  object        
 4   código unidade gestora     185211 non-null  float64       
 5   nome unidade gestora       191552 non-null  object        
 6   categoria econômica        192399 non-null  object        
 7   origem receita             186042 non-null  object        
 8   espécie receita            190129 non-null  object        
 9   detalhamento               186707 non-null  object        
 10  valor previsto atualizado  184806 non-null  float64       
 11  valor lançado              188960 non-null  float64 

In [99]:
df_17['nome órgão superior'] = df_17.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_17['nome órgão superior'].info() #Columna codigo y nombre


df_17["data lançamento"] = pd.to_datetime(df_17["data lançamento"]) 
df_17["data lançamento"]  #cambio a fecha

df_17["valor realizado"]= df_17["valor realizado"].str.replace(",",".")
df_17["valor realizado"] = list(map(float, df_17["valor realizado"]))
df_17["valor lançado"]= df_17["valor lançado"].str.replace(",",".")
df_17["valor lançado"] = list(map(float, df_17["valor lançado"]))
df_17["valor previsto atualizado"]= df_17["valor previsto atualizado"].str.replace(",",".")
df_17["valor previsto atualizado"] = list(map(float, df_17["valor previsto atualizado"]))
df_17["percentual realizado"]= df_17["percentual realizado"].str.replace(",",".")
df_17["percentual realizado"] = list(map(float, df_17["percentual realizado"]))    #cambio a float

C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\4205365010.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_17['nome órgão superior'] = df_17.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\4205365010.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_17["data lançamento"] = pd.to_datetime(df_17["data lançamento"])


<class 'pandas.core.series.Series'>
RangeIndex: 190479 entries, 0 to 190478
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
185979 non-null  object
dtypes: object(1)
memory usage: 1.5+ MB


In [100]:
df_17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190479 entries, 0 to 190478
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      185999 non-null  float64       
 1   nome órgão superior        185979 non-null  object        
 2   código órgão               188418 non-null  float64       
 3   nome órgão                 185457 non-null  object        
 4   código unidade gestora     186949 non-null  float64       
 5   nome unidade gestora       188076 non-null  object        
 6   categoria econômica        187950 non-null  object        
 7   origem receita             181808 non-null  object        
 8   espécie receita            181798 non-null  object        
 9   detalhamento               186726 non-null  object        
 10  valor previsto atualizado  180955 non-null  float64       
 11  valor lançado              182124 non-null  float64 

In [101]:
df_18['nome órgão superior'] = df_18.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_18['nome órgão superior'].info() #Columna codigo y nombre


df_18["data lançamento"] = pd.to_datetime(df_18["data lançamento"]) 
df_18["data lançamento"]  #cambio a fecha

df_18["valor realizado"]= df_18["valor realizado"].str.replace(",",".")
df_18["valor realizado"] = list(map(float, df_18["valor realizado"]))
df_18["valor lançado"]= df_18["valor lançado"].str.replace(",",".")
df_18["valor lançado"] = list(map(float, df_18["valor lançado"]))
df_18["valor previsto atualizado"]= df_18["valor previsto atualizado"].str.replace(",",".")
df_18["valor previsto atualizado"] = list(map(float, df_18["valor previsto atualizado"]))
df_18["percentual realizado"]= df_18["percentual realizado"].str.replace(",",".")
df_18["percentual realizado"] = list(map(float, df_18["percentual realizado"]))    #cambio a float

C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1489711203.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_18['nome órgão superior'] = df_18.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1489711203.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_18['nome órgão superior'] = df_18.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))


<class 'pandas.core.series.Series'>
RangeIndex: 173944 entries, 0 to 173943
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
165679 non-null  object
dtypes: object(1)
memory usage: 1.3+ MB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1489711203.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_18["data lançamento"] = pd.to_datetime(df_18["data lançamento"])


In [102]:
df_18.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173944 entries, 0 to 173943
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      165690 non-null  float64       
 1   nome órgão superior        165679 non-null  object        
 2   código órgão               168716 non-null  float64       
 3   nome órgão                 166611 non-null  object        
 4   código unidade gestora     166913 non-null  float64       
 5   nome unidade gestora       171411 non-null  object        
 6   categoria econômica        170516 non-null  object        
 7   origem receita             168037 non-null  object        
 8   espécie receita            166386 non-null  object        
 9   detalhamento               167435 non-null  object        
 10  valor previsto atualizado  165247 non-null  float64       
 11  valor lançado              170168 non-null  float64 

In [103]:
df_19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176828 entries, 0 to 176827
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   código órgão superior      170865 non-null  float64
 1   nome órgão superior        114938 non-null  object 
 2   código órgão               173643 non-null  float64
 3   nome órgão                 169862 non-null  object 
 4   código unidade gestora     171547 non-null  float64
 5   nome unidade gestora       174840 non-null  object 
 6   categoria econômica        174890 non-null  object 
 7   origem receita             168623 non-null  object 
 8   espécie receita            172708 non-null  object 
 9   detalhamento               172006 non-null  object 
 10  valor previsto atualizado  167987 non-null  object 
 11  valor lançado              172190 non-null  object 
 12  valor realizado            169191 non-null  object 
 13  percentual realizado       17

In [104]:
df_19['nome órgão superior'] = df_19.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_19['nome órgão superior'].info() #Columna codigo y nombre


df_19["data lançamento"] = pd.to_datetime(df_19["data lançamento"], format="%d/%m/%Y")
df_19["data lançamento"]  #cambio a fecha

df_19["valor realizado"]= df_19["valor realizado"].str.replace(",",".")
df_19["valor realizado"] = list(map(float, df_19["valor realizado"]))
df_19["valor lançado"]= df_19["valor lançado"].str.replace(",",".")
df_19["valor lançado"] = list(map(float, df_19["valor lançado"]))
df_19["valor previsto atualizado"]= df_19["valor previsto atualizado"].str.replace(",",".")
df_19["valor previsto atualizado"] = list(map(float, df_19["valor previsto atualizado"]))
df_19["percentual realizado"]= df_19["percentual realizado"].str.replace(",",".")
df_19["percentual realizado"] = list(map(float, df_19["percentual realizado"]))    #cambio a float

<class 'pandas.core.series.Series'>
RangeIndex: 176828 entries, 0 to 176827
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
170857 non-null  object
dtypes: object(1)
memory usage: 1.3+ MB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1939501729.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_19['nome órgão superior'] = df_19.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1939501729.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_19['nome órgão superior'] = df_19.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))


In [105]:
df_19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176828 entries, 0 to 176827
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      170865 non-null  float64       
 1   nome órgão superior        170857 non-null  object        
 2   código órgão               173643 non-null  float64       
 3   nome órgão                 169862 non-null  object        
 4   código unidade gestora     171547 non-null  float64       
 5   nome unidade gestora       174840 non-null  object        
 6   categoria econômica        174890 non-null  object        
 7   origem receita             168623 non-null  object        
 8   espécie receita            172708 non-null  object        
 9   detalhamento               172006 non-null  object        
 10  valor previsto atualizado  167987 non-null  float64       
 11  valor lançado              172190 non-null  float64 

In [106]:
df_20['nome órgão superior'] = df_20.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_20['nome órgão superior'].info() #Columna codigo y nombre


df_20["data lançamento"] = pd.to_datetime(df_20["data lançamento"]) 
df_20["data lançamento"]  #cambio a fecha

df_20["valor realizado"]= df_20["valor realizado"].str.replace(",",".")
df_20["valor realizado"] = list(map(float, df_20["valor realizado"]))
df_20["valor lançado"]= df_20["valor lançado"].str.replace(",",".")
df_20["valor lançado"] = list(map(float, df_20["valor lançado"]))
df_20["valor previsto atualizado"]= df_20["valor previsto atualizado"].str.replace(",",".")
df_20["valor previsto atualizado"] = list(map(float, df_20["valor previsto atualizado"]))
df_20["percentual realizado"]= df_20["percentual realizado"].str.replace(",",".")
df_20["percentual realizado"] = list(map(float, df_20["percentual realizado"]))    #cambio a float

C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\667092361.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_20['nome órgão superior'] = df_20.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\667092361.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df_20['nome órgão superior'] = df_20.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))


<class 'pandas.core.series.Series'>
RangeIndex: 142348 entries, 0 to 142347
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
137034 non-null  object
dtypes: object(1)
memory usage: 1.1+ MB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\667092361.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_20["data lançamento"] = pd.to_datetime(df_20["data lançamento"])


In [107]:
df_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142348 entries, 0 to 142347
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      137044 non-null  float64       
 1   nome órgão superior        137034 non-null  object        
 2   código órgão               135491 non-null  float64       
 3   nome órgão                 135255 non-null  object        
 4   código unidade gestora     138764 non-null  float64       
 5   nome unidade gestora       139702 non-null  object        
 6   categoria econômica        136452 non-null  object        
 7   origem receita             138019 non-null  object        
 8   espécie receita            137467 non-null  object        
 9   detalhamento               138373 non-null  object        
 10  valor previsto atualizado  135231 non-null  float64       
 11  valor lançado              140252 non-null  float64 

In [108]:



df_21['nome órgão superior'] = df_21.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_21['nome órgão superior'].info() #Columna codigo y nombre


df_21["data lançamento"] = pd.to_datetime(df_21["data lançamento"]) 
df_21["data lançamento"]  #cambio a fecha

df_21["valor realizado"]= df_21["valor realizado"].str.replace(",",".")
df_21["valor realizado"] = list(map(float, df_21["valor realizado"]))
df_21["valor lançado"]= df_21["valor lançado"].str.replace(",",".")
df_21["valor lançado"] = list(map(float, df_21["valor lançado"]))
df_21["valor previsto atualizado"]= df_21["valor previsto atualizado"].str.replace(",",".")
df_21["valor previsto atualizado"] = list(map(float, df_21["valor previsto atualizado"]))
df_21["percentual realizado"]= df_21["percentual realizado"].str.replace(",",".")
df_21["percentual realizado"] = list(map(float, df_21["percentual realizado"]))    #cambio a float



C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\3239938643.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_21['nome órgão superior'] = df_21.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))


<class 'pandas.core.series.Series'>
RangeIndex: 134593 entries, 0 to 134592
Series name: nome órgão superior
Non-Null Count   Dtype 
--------------   ----- 
133186 non-null  object
dtypes: object(1)
memory usage: 1.0+ MB


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\3239938643.py:5: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df_21["data lançamento"] = pd.to_datetime(df_21["data lançamento"])


In [109]:
df_21.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134593 entries, 0 to 134592
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      133191 non-null  float64       
 1   nome órgão superior        133186 non-null  object        
 2   código órgão               131202 non-null  float64       
 3   nome órgão                 129997 non-null  object        
 4   código unidade gestora     130352 non-null  float64       
 5   nome unidade gestora       128118 non-null  object        
 6   categoria econômica        131846 non-null  object        
 7   origem receita             132106 non-null  object        
 8   espécie receita            132754 non-null  object        
 9   detalhamento               132439 non-null  object        
 10  valor previsto atualizado  127863 non-null  float64       
 11  valor lançado              133148 non-null  float64 

In [110]:
df_13.head(3)    #el primer nan no soy capaz de eliminarlo pero como los he reducido bastante me doy con un canto en los dientes
                 #aunque creo que se podría utilizar el código de organización para rellenar los que quedan.

código órgão superior       nome órgão superior  código órgão  \
0                63000.0                       NaN       63000.0   
1                63000.0  Advocacia-Geral da União       63000.0   
2                63000.0  Advocacia-Geral da União       63000.0   

                                          nome órgão  código unidade gestora  \
0  Advocacia-Geral da União - Unidades com víncul...                110060.0   
1  Advocacia-Geral da União - Unidades com víncul...                110060.0   
2  Advocacia-Geral da União - Unidades com víncul...                110060.0   

                            nome unidade gestora categoria econômica  \
0  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
1  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   
2  COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU  Receitas Correntes   

              origem receita                                espécie receita  \
0  Outras Receitas Correntes  Bens, Direitos e Valores Incorporados ao Patr   
1  Outras Receitas Correntes    Indenizações, restituições e ressarcimentos   
2  Outras Receitas Correntes  Multas administrativas, contratuais e judicia   

                                    detalhamento  valor previsto atualizado  \
0  REC.DIVIDA ATIVA NAO TRIBUTARIA DE OUTRAS REC                        0.0   
1   RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES                        0.0   
2                  OUTRAS MULTAS E JUROS DE MORA                        0.0   

   valor lançado  valor realizado  percentual realizado data lançamento  \
0            0.0          1297.13                   0.0      2013-12-31   
1            0.0      26666621.42                   0.0      2013-12-31   
2            0.0        301251.13                   0.0      2013-12-31   

   ano exercício  
0            NaN  
1         2013.0  
2         2013.0

In [111]:
df_13.sample(3)

código órgão superior     nome órgão superior  código órgão  \
475                 52000.0    Ministério da Defesa       52221.0   
1598                26000.0  Ministério da Educação       26260.0   
1508                26000.0  Ministério da Educação       26238.0   

                                  nome órgão  código unidade gestora  \
475   Indústria de Material Bélico do Brasil                168002.0   
1598         Universidade Federal de Alfenas                153028.0   
1508    Universidade Federal de Minas Gerais                153062.0   

                               nome unidade gestora categoria econômica  \
475   INDUSTRIA DE MATERIAL BELICO DO BRASIL/GESTOR  Receitas Correntes   
1598                UNIVERSIDADE FEDERAL DE ALFENAS  Receitas Correntes   
1508           UNIVERSIDADE FEDERAL DE MINAS GERAIS  Receitas Correntes   

                 origem receita  \
475   Outras Receitas Correntes   
1598  Outras Receitas Correntes   
1508   Transferências Correntes   

                                    espécie receita  \
475                       Demais receitas correntes   
1598    Indenizações, restituições e ressarcimentos   
1508  Transferências dos Municípios e de suas Entid   

                                       detalhamento  \
475     REC.SEGUROS DECOR. INDENIZAÇÃO POR SINISTRO   
1598   RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   
1508  Transferências dos Municípios e de suas Entid   

      valor previsto atualizado  valor lançado  valor realizado  \
475                         0.0            NaN              NaN   
1598                        0.0            0.0         11539.14   
1508                   321141.0            0.0         20648.00   

      percentual realizado data lançamento  ano exercício  
475                    0.0      2013-12-31         2013.0  
1598                   0.0             NaT         2013.0  
1508                   6.0      2013-12-31         2013.0

**Describes**

In [112]:
df_13.describe(include ="O").T

count unique  \
nome órgão superior   4263     24   
nome órgão            4286    264   
nome unidade gestora  4278    279   
categoria econômica   4303      4   
origem receita        4413     15   
espécie receita       4355     44   
detalhamento          4377    694   

                                                                    top  freq  
nome órgão superior                              Ministério da Educação  1752  
nome órgão            Ministério da Economia - Unidades com vínculo ...   476  
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   392  
categoria econômica                                  Receitas Correntes  3829  
origem receita                                Outras Receitas Correntes  1945  
espécie receita                                     Receita de Serviços   955  
detalhamento               RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   210

In [113]:
df_13.describe().T

count                 mean                  min  \
código órgão superior      4282.0          31739.60766              20000.0   
código órgão               4362.0         31484.209995              20101.0   
código unidade gestora     4324.0        232766.252775             110005.0   
valor previsto atualizado  4273.0     498180092.502467                  0.0   
valor lançado              4282.0     265722648.433828                  0.0   
valor realizado            4309.0     386036258.037721        -276398836.04   
percentual realizado       4360.0           1025.67156               -201.0   
data lançamento              4398  2013-12-31 00:00:00  2013-12-31 00:00:00   
ano exercício              3374.0               2013.0               2013.0   

                                           25%                  50%  \
código órgão superior                  25000.0              26000.0   
código órgão                           26000.0              26292.0   
código unidade gestora                153978.0             160075.0   
valor previsto atualizado                  0.0              23428.0   
valor lançado                              0.0                  0.0   
valor realizado                         505.69             48211.14   
percentual realizado                       0.0                  0.0   
data lançamento            2013-12-31 00:00:00  2013-12-31 00:00:00   
ano exercício                           2013.0               2013.0   

                                           75%                  max  \
código órgão superior                  36000.0              81000.0   
código órgão                           35000.0              91214.0   
código unidade gestora                253003.0             913001.0   
valor previsto atualizado            1043267.0       610065700590.0   
valor lançado                              0.0  295606275807.679993   
valor realizado                      978909.54  413786247189.789978   
percentual realizado                     98.25            1605449.0   
data lançamento            2013-12-31 00:00:00  2013-12-31 00:00:00   
ano exercício                           2013.0               2013.0   

                                         std  
código órgão superior           10692.607398  
código órgão                    10810.318732  
código unidade gestora         141238.719759  
valor previsto atualizado  10519744250.78565  
valor lançado              6212270728.869926  
valor realizado            7247324154.900342  
percentual realizado            26471.314837  
data lançamento                          NaN  
ano exercício                            0.0

In [114]:
df_14.describe().T                          #como va a haber porcentajes negativos

count                 mean                  min  \
código órgão superior      4331.0         31847.841145              20000.0   
código órgão               4493.0         31818.443356              20101.0   
código unidade gestora     4332.0        236574.978532             110005.0   
valor previsto atualizado  4325.0     539445992.952192                  0.0   
valor lançado              4375.0     287561053.098779                  0.0   
valor realizado            4328.0     510978230.763729        -624211482.24   
percentual realizado       4437.0          17200.12238            -114552.0   
data lançamento              4411  2014-12-31 00:00:00  2014-12-31 00:00:00   
ano exercício              3415.0               2014.0               2014.0   

                                           25%                  50%  \
código órgão superior                  25000.0              26000.0   
código órgão                           26000.0              26298.0   
código unidade gestora                153167.0             160075.0   
valor previsto atualizado                  0.0              17890.0   
valor lançado                              0.0                  0.0   
valor realizado                       336.7475             43004.28   
percentual realizado                       0.0                  0.0   
data lançamento            2014-12-31 00:00:00  2014-12-31 00:00:00   
ano exercício                           2014.0               2014.0   

                                           75%                  max  \
código órgão superior                  36000.0              81000.0   
código órgão                           36201.0              91214.0   
código unidade gestora                255000.0             913001.0   
valor previsto atualizado            1281585.0       654746947069.0   
valor lançado                              0.0  357160677863.969971   
valor realizado                       961652.9  548491208660.309998   
percentual realizado                      95.0           72363772.0   
data lançamento            2014-12-31 00:00:00  2014-12-31 00:00:00   
ano exercício                           2014.0               2014.0   

                                          std  
código órgão superior            10813.244886  
código órgão                     11143.198081  
código unidade gestora          145862.242362  
valor previsto atualizado  11348956512.135315  
valor lançado               7168037787.526263  
valor realizado            10314230981.388567  
percentual realizado           1086744.364682  
data lançamento                           NaN  
ano exercício                             0.0

In [115]:
df_14.describe(include= "O").T

count unique  \
nome órgão superior   4325     25   
nome órgão            4439    267   
nome unidade gestora  4392    278   
categoria econômica   4502      5   
origem receita        4487     14   
espécie receita       4455     46   
detalhamento          4506    706   

                                                                    top  freq  
nome órgão superior                              Ministério da Educação  1747  
nome órgão            Ministério da Economia - Unidades com vínculo ...   465  
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   365  
categoria econômica                                  Receitas Correntes  4051  
origem receita                                Outras Receitas Correntes  2025  
espécie receita           Multas administrativas, contratuais e judicia   951  
detalhamento               RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   214

In [116]:
df_15.describe().T           #como va a haber porcentajes negativos

count                 mean                  min  \
código órgão superior      4324.0         31970.860315              20000.0   
código órgão               4351.0         32019.609975              20101.0   
código unidade gestora     4326.0         237160.81877             110005.0   
valor previsto atualizado  4297.0     668178811.234284                  0.0   
valor lançado              4381.0      69923679.842525                  0.0   
valor realizado            4351.0     605428383.607584         -500851791.0   
percentual realizado       4322.0           348.894493               -402.0   
data lançamento              4398  2015-12-31 00:00:00  2015-12-31 00:00:00   
ano exercício              3392.0               2015.0               2015.0   

                                           25%                  50%  \
código órgão superior                  25000.0              26000.0   
código órgão                           26000.0              26298.0   
código unidade gestora                153173.0             167086.0   
valor previsto atualizado                  0.0              22284.0   
valor lançado                              0.0                  0.0   
valor realizado                         393.68             44651.79   
percentual realizado                       0.0                  0.0   
data lançamento            2015-12-31 00:00:00  2015-12-31 00:00:00   
ano exercício                           2015.0               2015.0   

                                           75%                  max  \
código órgão superior                  36000.0              81000.0   
código órgão                           36211.0              91214.0   
código unidade gestora                264001.0             913001.0   
valor previsto atualizado            1641495.0       902028908307.0   
valor lançado                              0.0   54211937114.779999   
valor realizado                     914492.185  771117711060.949951   
percentual realizado                     80.75             745465.0   
data lançamento            2015-12-31 00:00:00  2015-12-31 00:00:00   
ano exercício                           2015.0               2015.0   

                                          std  
código órgão superior            10941.671616  
código órgão                     11465.439048  
código unidade gestora          145720.161158  
valor previsto atualizado  15204410374.208685  
valor lançado               1365041936.360889  
valor realizado            13297834078.761972  
percentual realizado             11562.793987  
data lançamento                           NaN  
ano exercício                             0.0

In [117]:
df_15.describe(include= "O").T

count unique  \
nome órgão superior   4308     25   
nome órgão            4361    271   
nome unidade gestora  4449    288   
categoria econômica   4463      4   
origem receita        4346     14   
espécie receita       4320     41   
detalhamento          4393    694   

                                                                    top  freq  
nome órgão superior                              Ministério da Educação  1721  
nome órgão            Ministério da Economia - Unidades com vínculo ...   466  
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   374  
categoria econômica                                  Receitas Correntes  4037  
origem receita                                Outras Receitas Correntes  1999  
espécie receita           Multas administrativas, contratuais e judicia   927  
detalhamento               RECUPERACAO DE DESPESAS DE EXERC. ANTERIORES   218

In [118]:
df_16.describe().T   #como va a haber porcentajes negativos

count                           mean  \
código órgão superior      190214.0                   32735.986836   
código órgão               190506.0                   32087.913158   
código unidade gestora     185211.0                  231662.563579   
valor previsto atualizado  184806.0                15747273.497974   
valor lançado              188960.0                 1754045.674195   
valor realizado            188766.0                14765270.649226   
percentual realizado       192008.0                       0.008239   
data lançamento              192088  2016-06-29 04:23:03.266002944   
ano exercício              145900.0                         2016.0   

                                           min                  25%  \
código órgão superior                  20000.0              26000.0   
código órgão                           20101.0              26232.0   
código unidade gestora                110005.0             153165.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado              -54211937114.779999                  0.0   
valor realizado           -156285934188.029999                200.0   
percentual realizado                      -3.0                  0.0   
data lançamento            2016-01-01 00:00:00  2016-03-30 00:00:00   
ano exercício                           2016.0               2016.0   

                                           50%                  75%  \
código órgão superior                  26000.0              39000.0   
código órgão                           26406.0              36212.0   
código unidade gestora                167086.0             253003.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                         2060.0            27411.125   
percentual realizado                       0.0                  0.0   
data lançamento            2016-06-29 00:00:00  2016-09-29 00:00:00   
ano exercício                           2016.0               2016.0   

                                           max                std  
código órgão superior                  81000.0       11072.787723  
código órgão                           91214.0       10958.211896  
código unidade gestora                913001.0      143282.285279  
valor previsto atualizado       883028610465.0  2430183666.093293  
valor lançado               93962415840.259995   443967664.829929  
valor realizado            199853249683.029999   892743921.484681  
percentual realizado                     385.0            1.23814  
data lançamento            2016-12-31 00:00:00                NaN  
ano exercício                           2016.0                0.0

In [119]:
df_16.describe(include= "O").T

count unique  \
nome órgão superior   190203     25   
nome órgão            191144    269   
nome unidade gestora  191552    292   
categoria econômica   192399      4   
origem receita        186042     14   
espécie receita       190129     44   
detalhamento          186707    525   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                Outras Receitas Correntes   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    70266  
nome órgão             13510  
nome unidade gestora   10952  
categoria econômica   184635  
origem receita         67579  
espécie receita        53659  
detalhamento           28989

In [120]:
df_17.describe().T

count                           mean  \
código órgão superior      185999.0                   32547.115845   
código órgão               188418.0                   31814.245996   
código unidade gestora     186949.0                   231583.87488   
valor previsto atualizado  180955.0                 18792979.68145   
valor lançado              182124.0                 2490210.347853   
valor realizado            182807.0                13544741.498324   
percentual realizado       186281.0                       0.010189   
data lançamento              183488  2017-06-29 21:11:18.060690688   
ano exercício              142859.0                         2017.0   

                                           min                  25%  \
código órgão superior                  20000.0              26000.0   
código órgão                           20101.0              26231.0   
código unidade gestora                110005.0             153166.0   
valor previsto atualizado         -214773743.0                  0.0   
valor lançado                  -12957230886.98                  0.0   
valor realizado            -24922519115.560001                216.0   
percentual realizado                       0.0                  0.0   
data lançamento            2017-01-01 00:00:00  2017-03-30 00:00:00   
ano exercício                           2017.0               2017.0   

                                           50%                  75%  \
código órgão superior                  26000.0              39000.0   
código órgão                           26405.0              36201.0   
código unidade gestora                167086.0             253003.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        2247.84            29592.885   
percentual realizado                       0.0                  0.0   
data lançamento            2017-06-30 00:00:00  2017-09-28 00:00:00   
ano exercício                           2017.0               2017.0   

                                           max                std  
código órgão superior                  81000.0       10955.975665  
código órgão                           91214.0       10681.942444  
código unidade gestora                913001.0      142775.277196  
valor previsto atualizado       925084468997.0  2862904490.583281  
valor lançado               84511925102.119995   289931323.768856  
valor realizado            100717273794.669998   435040858.480846  
percentual realizado                     742.0           1.867604  
data lançamento            2017-12-31 00:00:00                NaN  
ano exercício                           2017.0                0.0

In [121]:
df_17.describe(include= "O").T

count unique  \
nome órgão superior   185979     24   
nome órgão            185457    268   
nome unidade gestora  188076    318   
categoria econômica   187950      5   
origem receita        181808     15   
espécie receita       181798     47   
detalhamento          186726    514   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                Outras Receitas Correntes   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    69137  
nome órgão             13789  
nome unidade gestora   11063  
categoria econômica   180236  
origem receita         65611  
espécie receita        49265  
detalhamento           29255

In [122]:
df_18.describe().T

count                           mean  \
código órgão superior      165690.0                   31425.408896   
código órgão               168716.0                   30759.630468   
código unidade gestora     166913.0                  224882.574233   
valor previsto atualizado  165247.0                21039020.990555   
valor lançado              170168.0                 2754248.766615   
valor realizado            165640.0                17302446.191829   
percentual realizado       171264.0                       0.001734   
data lançamento              169298  2018-06-29 16:29:38.884570624   
ano exercício              130458.0                         2018.0   

                                           min                  25%  \
código órgão superior                  20000.0              25000.0   
código órgão                           20101.0              25000.0   
código unidade gestora                110005.0             154034.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado              -10368360188.309999                  0.0   
valor realizado                  -3500000000.0              295.485   
percentual realizado                       0.0                  0.0   
data lançamento            2018-01-01 00:00:00  2018-03-30 00:00:00   
ano exercício                           2018.0               2018.0   

                                           50%                  75%  \
código órgão superior                  26000.0              36000.0   
código órgão                           26285.0              32396.0   
código unidade gestora                167086.0             240901.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        3036.24             37590.66   
percentual realizado                       0.0                  0.0   
data lançamento            2018-06-29 00:00:00  2018-09-28 00:00:00   
ano exercício                           2018.0               2018.0   

                                           max                std  
código órgão superior                  81000.0       10166.459924  
código órgão                           91214.0        9821.452882  
código unidade gestora                913001.0      138519.077341  
valor previsto atualizado      1130479484929.0  3294057493.632579  
valor lançado              125769983134.070007   395273184.970806  
valor realizado            169256367162.369995   632774794.485745  
percentual realizado                      36.0           0.139409  
data lançamento            2018-12-31 00:00:00                NaN  
ano exercício                           2018.0                0.0

In [123]:
df_18.describe(include= "O").T

count unique  \
nome órgão superior   165679     20   
nome órgão            166611    260   
nome unidade gestora  171411    312   
categoria econômica   170516      5   
origem receita        168037     15   
espécie receita       166386     47   
detalhamento          167435    608   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                      Receita de Serviços   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    62790  
nome órgão             21908  
nome unidade gestora   20882  
categoria econômica   162326  
origem receita         56686  
espécie receita        49206  
detalhamento           28215

In [124]:
df_19.describe().T   #como va a haber porcentajes negativos

count                           mean  \
código órgão superior      170865.0                   31409.703567   
código órgão               173643.0                   30825.272945   
código unidade gestora     171547.0                  225092.502475   
valor previsto atualizado  167987.0                19341842.193962   
valor lançado              172190.0                 1971936.871204   
valor realizado            169191.0                17024688.477692   
percentual realizado       170075.0                       0.186912   
data lançamento              173914  2019-07-01 17:13:10.859849984   
ano exercício              132621.0                         2019.0   

                                           min                  25%  \
código órgão superior                  20000.0              25000.0   
código órgão                           20101.0              25000.0   
código unidade gestora                110005.0             154039.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado             -148347267246.910004                  0.0   
valor realizado            -60439234089.120003                315.0   
percentual realizado                    -282.0                  0.0   
data lançamento            2019-01-01 00:00:00  2019-04-02 00:00:00   
ano exercício                           2019.0               2019.0   

                                           50%                  75%  \
código órgão superior                  26000.0              36000.0   
código órgão                           26290.0              32396.0   
código unidade gestora                167086.0             240102.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        3262.66             39379.27   
percentual realizado                       0.0                  0.0   
data lançamento            2019-07-03 00:00:00  2019-09-30 00:00:00   
ano exercício                           2019.0               2019.0   

                                           max                std  
código órgão superior                  81000.0       10161.744743  
código órgão                           91214.0        9834.693796  
código unidade gestora                913001.0      139752.985581  
valor previsto atualizado       747311062387.0  2512352098.437716  
valor lançado              146326992731.609985   662442988.866594  
valor realizado             50193922075.790001    479651393.25194  
percentual realizado                   16631.0          49.095948  
data lançamento            2019-12-31 00:00:00                NaN  
ano exercício                           2019.0                0.0

In [125]:
df_19.describe(include= "O").T

count unique  \
nome órgão superior   170857     19   
nome órgão            169862    248   
nome unidade gestora  174840    269   
categoria econômica   174890      5   
origem receita        168623     15   
espécie receita       172708     46   
detalhamento          172006    635   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                      Receita de Serviços   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    64798  
nome órgão             22593  
nome unidade gestora   22102  
categoria econômica   167342  
origem receita         57171  
espécie receita        51505  
detalhamento           28394

In [126]:
df_20.describe().T

count                           mean  \
código órgão superior      137044.0                   32086.125624   
código órgão               135491.0                   31579.852618   
código unidade gestora     138764.0                  238443.493471   
valor previsto atualizado  135231.0                26090112.829168   
valor lançado              140252.0                 4447722.713776   
valor realizado            138063.0                25064651.825275   
percentual realizado       140314.0                       0.285909   
data lançamento              139954  2020-06-25 08:48:26.257770752   
ano exercício              106761.0                         2020.0   

                                           min                  25%  \
código órgão superior                  20000.0              25000.0   
código órgão                           20101.0              25000.0   
código unidade gestora                110005.0             154044.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado              -83137328053.639999                  0.0   
valor realizado                -22288681258.25              317.015   
percentual realizado                       0.0                  0.0   
data lançamento            2020-01-01 00:00:00  2020-03-17 00:00:00   
ano exercício                           2020.0               2020.0   

                                           50%                  75%  \
código órgão superior                  26000.0              39000.0   
código órgão                           26290.0              36201.0   
código unidade gestora                170013.0             253003.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        4346.98             58287.61   
percentual realizado                       0.0                  0.0   
data lançamento            2020-06-26 00:00:00  2020-09-29 00:00:00   
ano exercício                           2020.0               2020.0   

                                           max                std  
código órgão superior                  81000.0        10598.32121  
código órgão                           90000.0       10454.175702  
código unidade gestora                873001.0      151544.696762  
valor previsto atualizado       915650943341.0  3151089045.615634  
valor lançado              257513407925.600006   812872097.657915  
valor realizado            350022251304.299988  1113697027.941124  
percentual realizado                   39232.0         104.734931  
data lançamento            2020-12-30 00:00:00                NaN  
ano exercício                           2020.0                0.0

In [127]:
df_20.describe(include= "O").T

count unique  \
nome órgão superior   137034     19   
nome órgão            135255    253   
nome unidade gestora  139702    268   
categoria econômica   136452      5   
origem receita        138019     15   
espécie receita       137467     45   
detalhamento          138373    624   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                Outras Receitas Correntes   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    39339  
nome órgão             21549  
nome unidade gestora   20696  
categoria econômica   129594  
origem receita         40864  
espécie receita        32138  
detalhamento           20397

In [128]:
df_21.describe().T #como va a haber porcentajes negativos

count                           mean  \
código órgão superior      133191.0                   31980.689386   
código órgão               131202.0                    31457.17534   
código unidade gestora     130352.0                  238690.054506   
valor previsto atualizado  127863.0                31622589.556976   
valor lançado              133148.0                 2680547.612767   
valor realizado            129379.0                28436292.079754   
percentual realizado       129104.0                       0.002541   
data lançamento              130519  2021-06-16 01:38:51.274373888   
ano exercício              100945.0                         2021.0   

                                           min                  25%  \
código órgão superior                  20000.0              25000.0   
código órgão                           20101.0              25000.0   
código unidade gestora                110005.0             154044.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                  -37505821629.82                  0.0   
valor realizado            -12126486130.059999               306.17   
percentual realizado                      -1.0                  0.0   
data lançamento            2021-01-01 00:00:00  2021-03-26 00:00:00   
ano exercício                           2021.0               2021.0   

                                           50%                  75%  \
código órgão superior                  26000.0              39000.0   
código órgão                           26285.0              36000.0   
código unidade gestora                170013.0             253002.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        4350.24            60616.345   
percentual realizado                       0.0                  0.0   
data lançamento            2021-06-15 00:00:00  2021-09-06 00:00:00   
ano exercício                           2021.0               2021.0   

                                           max                std  
código órgão superior                  81000.0        10477.36015  
código órgão                           90000.0       10274.244983  
código unidade gestora                873001.0      151288.080629  
valor previsto atualizado      1603521711208.0  4846032058.977045  
valor lançado              114112106824.449997   400887058.951995  
valor realizado             55340292367.190002   703122327.495834  
percentual realizado                      51.0           0.238297  
data lançamento            2021-12-03 00:00:00                NaN  
ano exercício                           2021.0                0.0

In [129]:
df_21.describe(include= "O").T

count unique  \
nome órgão superior   133186     20   
nome órgão            129997    256   
nome unidade gestora  128118    267   
categoria econômica   131846      5   
origem receita        132106     15   
espécie receita       132754     42   
detalhamento          132439    598   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                      Receita de Serviços   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior    39453  
nome órgão             20681  
nome unidade gestora   19344  
categoria econômica   125477  
origem receita         39200  
espécie receita        33667  
detalhamento           19208

*Para rellenar el año que falta porque pone nan*

In [130]:
listadf= [df_13,df_14,df_15,df_16,df_17,df_18,df_19,df_20,df_21]
lista_anio= [2013,2014,2015,2016,2016,2017,2018,2019,2020,2021]
def fillyear(df, year):
    df["ano exercício"] = df["ano exercício"].fillna(year)
    return df

In [131]:
listadf_aniosbien = list(map(fillyear, listadf, lista_anio))

In [132]:
df_19.info()    #Comprobación de que hemos acabado con los nulos en año lanzamiento y reducido los nulos de nombre órgano superior

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176828 entries, 0 to 176827
Data columns (total 16 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   código órgão superior      170865 non-null  float64       
 1   nome órgão superior        170857 non-null  object        
 2   código órgão               173643 non-null  float64       
 3   nome órgão                 169862 non-null  object        
 4   código unidade gestora     171547 non-null  float64       
 5   nome unidade gestora       174840 non-null  object        
 6   categoria econômica        174890 non-null  object        
 7   origem receita             168623 non-null  object        
 8   espécie receita            172708 non-null  object        
 9   detalhamento               172006 non-null  object        
 10  valor previsto atualizado  167987 non-null  float64       
 11  valor lançado              172190 non-null  float64 

**Mergeo**

In [133]:
df_completa = pd.concat(listadf_aniosbien).reset_index(drop= True)

completo nan en nome orgao y codigo unidade gestora (el nome orgao superior ya lo hice en las individuales)

In [134]:
df_completa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   código órgão superior      995940 non-null   float64       
 1   nome órgão superior        995834 non-null   object        
 2   código órgão               1001182 non-null  float64       
 3   nome órgão                 991412 non-null   object        
 4   código unidade gestora     992718 non-null   float64       
 5   nome unidade gestora       1006818 non-null  object        
 6   categoria econômica        1007321 non-null  object        
 7   origem receita             987881 non-null   object        
 8   espécie receita            994372 non-null   object        
 9   detalhamento               996962 non-null   object        
 10  valor previsto atualizado  974984 non-null   float64       
 11  valor lançado              999880 non

In [135]:
df_completa['nome órgão'] = df_completa.groupby('código órgão')['nome órgão'].transform(lambda x: x.fillna(method='ffill'))

df_completa['código unidade gestora'] = df_completa.groupby('nome unidade gestora')['código unidade gestora'].transform(lambda x: x.fillna(method='ffill'))

df_completa['nome órgão superior'] = df_completa.groupby('código órgão superior')['nome órgão superior'].transform(lambda x: x.fillna(method='ffill'))
df_completa.info() #Columna codigo y nombre


C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1560036237.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_completa['nome órgão'] = df_completa.groupby('código órgão')['nome órgão'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1560036237.py:3: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_completa['código unidade gestora'] = df_completa.groupby('nome unidade gestora')['código unidade gestora'].transform(lambda x: x.fillna(method='ffill'))
C:\Users\jaime\AppData\Local\Temp\ipykernel_24592\1560036237.py:5: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_completa['nome órgão superior'] = df_completa.groupby('código órgão superior')['nome órgão superior'].transform

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026299 entries, 0 to 1026298
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   código órgão superior      995940 non-null   float64       
 1   nome órgão superior        995921 non-null   object        
 2   código órgão               1001182 non-null  float64       
 3   nome órgão                 1001164 non-null  object        
 4   código unidade gestora     1006806 non-null  float64       
 5   nome unidade gestora       1006818 non-null  object        
 6   categoria econômica        1007321 non-null  object        
 7   origem receita             987881 non-null   object        
 8   espécie receita            994372 non-null   object        
 9   detalhamento               996962 non-null   object        
 10  valor previsto atualizado  974984 non-null   float64       
 11  valor lançado              999880 non

In [136]:
df_completa.shape

(1026299, 16)

**Duplicados para eliminar**

In [137]:
df_completa.duplicated().sum()     #Hay 73 duplicadas que habrá que eliminar

np.int64(160)

In [138]:
df_completa = df_completa.drop_duplicates()
df_completa.duplicated().sum()

np.int64(0)

In [139]:
df_completa.shape

(1026139, 16)

In [140]:
df_completa= df_completa.reset_index(drop= True)
df_completa.sample(3)

código órgão superior       nome órgão superior  código órgão  \
282130                26000.0    Ministério da Educação       26241.0   
947419                25000.0    Ministério da Economia       25208.0   
207142                20000.0  Presidência da República       20101.0   

                                  nome órgão  código unidade gestora  \
282130        Universidade Federal do Paraná                153079.0   
947419  Superintendência de Seguros Privados                173039.0   
207142              Presidência da República                110005.0   

                                 nome unidade gestora categoria econômica  \
282130                 UNIVERSIDADE FEDERAL DO PARANA  Receitas Correntes   
947419      SUPERINTENDENCIA DE SEGUROS PRIVADOS - RJ                 NaN   
207142  SECRETARIA DE ADMINISTRACAO-SET.ORC.E FIN./PR  Receitas Correntes   

                   origem receita  \
282130        Receita de Serviços   
947419  Outras Receitas Correntes   
207142  Outras Receitas Correntes   

                                      espécie receita  \
282130                                Outros serviços   
947419  Multas administrativas, contratuais e judicia   
207142    Indenizações, restituições e ressarcimentos   

                                         detalhamento  \
282130                      OUTROS SERVICOS-PRINCIPAL   
947419  MULTAS PREVISTAS EM LEGISL.ESPECIFICA-DIV.AT.   
207142                OUTROS RESSARCIMENTOS-PRINCIPAL   

        valor previsto atualizado  valor lançado  valor realizado  \
282130                        0.0            0.0          3310.89   
947419                        0.0            0.0          1881.63   
207142                        0.0            0.0             5.56   

        percentual realizado data lançamento  ano exercício  
282130                   0.0      2017-09-27         2017.0  
947419                   0.0      2021-02-24         2021.0  
207142                   0.0      2016-06-16         2016.0

In [142]:
df_completa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026139 entries, 0 to 1026138
Data columns (total 16 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   código órgão superior      995780 non-null   float64       
 1   nome órgão superior        995761 non-null   object        
 2   código órgão               1001022 non-null  float64       
 3   nome órgão                 1001004 non-null  object        
 4   código unidade gestora     1006646 non-null  float64       
 5   nome unidade gestora       1006658 non-null  object        
 6   categoria econômica        1007161 non-null  object        
 7   origem receita             987721 non-null   object        
 8   espécie receita            994212 non-null   object        
 9   detalhamento               996804 non-null   object        
 10  valor previsto atualizado  974824 non-null   float64       
 11  valor lançado              999720 non

In [143]:
df_completa.describe().T

count                           mean  \
código órgão superior       995780.0                   32053.627307   
código órgão               1001022.0                   31438.394377   
código unidade gestora     1006646.0                  231236.370269   
valor previsto atualizado   974824.0                28662846.250981   
valor lançado               999720.0                 5277628.195648   
valor realizado             986720.0                24990249.025955   
percentual realizado       1002005.0                      82.207977   
data lançamento              1002467  2018-09-20 04:04:29.177538816   
ano exercício              1026139.0                     2018.02629   

                                           min                  25%  \
código órgão superior                  20000.0              25000.0   
código órgão                           20101.0              25201.0   
código unidade gestora                110005.0             154003.0   
valor previsto atualizado         -214773743.0                  0.0   
valor lançado             -148347267246.910004                  0.0   
valor realizado           -156285934188.029999             260.9175   
percentual realizado                 -114552.0                  0.0   
data lançamento            2013-12-31 00:00:00  2017-03-30 00:00:00   
ano exercício                           2013.0               2017.0   

                                           50%                  75%  \
código órgão superior                  26000.0              39000.0   
código órgão                           26298.0              36000.0   
código unidade gestora                170013.0             250088.0   
valor previsto atualizado                  0.0                  0.0   
valor lançado                              0.0                  0.0   
valor realizado                        3028.83           40954.6975   
percentual realizado                       0.0                  0.0   
data lançamento            2018-08-29 00:00:00  2020-02-12 00:00:00   
ano exercício                           2018.0               2019.0   

                                           max                std  
código órgão superior                  81000.0       10615.007996  
código órgão                           91214.0       10386.325916  
código unidade gestora                913001.0      144125.698807  
valor previsto atualizado      1603521711208.0  3479204745.188661  
valor lançado              357160677863.969971   817154305.863975  
valor realizado            771117711060.949951  1418143773.186865  
percentual realizado                72363772.0       72342.466154  
data lançamento            2021-12-03 00:00:00                NaN  
ano exercício                           2021.0           1.715024

In [144]:
df_completa["valor previsto atualizado"].sum()

np.float64(27941230433766.598)

In [145]:
df_completa["valor lançado"].sum()

np.float64(5276150459753.68)

In [146]:
df_completa.describe(include= "O").T

count unique  \
nome órgão superior    995761     25   
nome órgão            1001004    287   
nome unidade gestora  1006658    356   
categoria econômica   1007161      5   
origem receita         987721     15   
espécie receita        994212     63   
detalhamento           996804   1883   

                                                                    top  \
nome órgão superior                              Ministério da Educação   
nome órgão            Ministério da Economia - Unidades com vínculo ...   
nome unidade gestora            SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
categoria econômica                                  Receitas Correntes   
origem receita                                Outras Receitas Correntes   
espécie receita            Serviços Administrativos e Comerciais Gerais   
detalhamento                SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

                        freq  
nome órgão superior   350908  
nome órgão            116581  
nome unidade gestora  106163  
categoria econômica   961372  
origem receita        310594  
espécie receita       269344  
detalhamento          154410

In [148]:
df_completa.sample(3)    #categoria economica podemos rellenar sus valores nulos si en origen lleva la palabra servicios, correntes, capital

código órgão superior     nome órgão superior  código órgão  \
505658                26000.0  Ministério da Educação       26428.0   
778646                25000.0  Ministério da Economia       25203.0   
149323                26000.0  Ministério da Educação       26271.0   

                               nome órgão  código unidade gestora  \
505658      Instituto Federal de Brasília                158143.0   
778646    Comissão de Valores Mobiliários                173030.0   
149323  Fundação Universidade de Brasília                154040.0   

                           nome unidade gestora categoria econômica  \
505658             IFB - REITORIA - BRASILIA-DF  Receitas Correntes   
778646          COMISSAO DE VALORES MOBILIARIOS  Receitas Correntes   
149323  FUNDACAO UNIVERSIDADE DE BRASILIA - FUB  Receitas Correntes   

                                     origem receita  \
505658                    Outras Receitas Correntes   
778646  Impostos, Taxas e Contribuições de Melhoria   
149323                    Outras Receitas Correntes   

                                    espécie receita  \
505658                    Demais receitas correntes   
778646                                        Taxas   
149323  Indenizações, restituições e ressarcimentos   

                                         detalhamento  \
505658       OUTRAS RECEITAS-PRIMARIAS-MULTAS E JUROS   
778646  TAXAS INSPEÇÃO,CONTROLE E FISCALIZAÇÃO-DÍV.AT   
149323                OUTROS RESSARCIMENTOS-PRINCIPAL   

        valor previsto atualizado  valor lançado  valor realizado  \
505658                        0.0            0.0            17.50   
778646                        0.0            0.0         26819.60   
149323                        0.0            0.0            21.85   

        percentual realizado data lançamento  ano exercício  
505658                   0.0      2018-10-05         2018.0  
778646                   NaN      2020-09-08         2019.0  
149323                   0.0      2016-01-22         2016.0

In [150]:
df_completa["categoria econômica"].describe(include="O")

count                1007161
unique                     5
top       Receitas Correntes
freq                  961372
Name: categoria econômica, dtype: object

In [151]:
dfinteresante= df_completa[["nome órgão superior","nome unidade gestora","categoria econômica","origem receita","espécie receita","detalhamento","valor previsto atualizado","valor lançado","valor realizado","percentual realizado","ano exercício"]]
dfinteresante.sample(3)

nome órgão superior                           nome unidade gestora  \
51935   Ministério da Economia        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
273575  Ministério da Economia                                            NaN   
678902                     NaN  FUND. UNIVERSIDADE FEDERAL VALE SAO FRANCISCO   

        categoria econômica        origem receita  \
51935   Receitas de Capital  Operações de Crédito   
273575   Receitas Correntes         Contribuições   
678902   Receitas Correntes   Receita Patrimonial   

                                      espécie receita  \
51935          Operações de crédito - mercado externo   
273575                          Contribuições sociais   
678902  Exploração do patrimônio imobiliário do Estad   

                                         detalhamento  \
51935   OP.DE CREDITO CONTRATUAIS-MERC.EXTERNO-PRINC.   
273575  CONTRIBUIÇÃO SOBRE LOTERIAS ESPORTIVAS-PRINC.   
678902             ALUGUEIS E ARRENDAMENTOS-PRINCIPAL   

        valor previsto atualizado  valor lançado  valor realizado  \
51935                         0.0            0.0       2377634.23   
273575                        0.0            0.0       1125976.43   
678902                        0.0            0.0           841.79   

        percentual realizado  ano exercício  
51935                    0.0         2016.0  
273575                   0.0         2017.0  
678902                   0.0         2019.0

**Corregir Categorías(Dentro de nulos)**

In [177]:
a=df_completa["detalhamento"] == "Sem informação"
a.sum()

np.int64(250)

In [155]:
dfprueba= df_completa.groupby("categoria econômica")[["valor previsto atualizado"]].sum()   #Esto es solo para visualizar
dfprueba

valor previsto atualizado
categoria econômica                                                 
Receitas Correntes                                      1.314475e+13
Receitas Correntes - intra-orçamentárias                2.694278e+11
Receitas de Capital                                     1.349986e+13
Receitas de Capital - intra-orçamentárias               1.080841e+10
Sem informação                                          0.000000e+00

Tengo que juntar nan y sin información en la misma

In [ ]:
df_completa["categoria econômica"].value_counts()

categoria econômica
Receitas Correntes                           961372
Receitas de Capital                           29521
Receitas Correntes - intra-orçamentárias      15926
Sem informação                                  252
Receitas de Capital - intra-orçamentárias        90
Name: count, dtype: int64

In [ ]:
df_completa["categoria econômica"].isna().value_counts()

categoria econômica
False    1007161
True       18978
Name: count, dtype: int64

In [ ]:
df_completa["categoria econômica"].isna().sum()

np.int64(18978)

**Fusión de nan y sem informacion**  

En los codigos, 0000 es el codigo que hemos puesto para los sin informacion (porque no se podía poner el str)

In [198]:
df_completa["categoria econômica"]= df_completa["categoria econômica"].fillna(value= "Sem informação")
df_completa["origem receita"]= df_completa["origem receita"].fillna(value= "Sem informação")
df_completa["espécie receita"]= df_completa["espécie receita"].fillna(value= "Sem informação")
df_completa["detalhamento"]= df_completa["detalhamento"].fillna(value= "Sem informação")

df_completa["nome órgão superior"]= df_completa["nome órgão superior"].fillna(value= "Sem informação")
df_completa["nome órgão"]= df_completa["nome órgão"].fillna(value= "Sem informação")
df_completa["nome unidade gestora"]= df_completa["nome unidade gestora"].fillna(value= "Sem informação")

df_completa["código órgão superior"]= df_completa["código órgão superior"].fillna(value= 0000)
df_completa["código órgão"]= df_completa["código órgão"].fillna(value= 0000)
df_completa["código unidade gestora"]= df_completa["código unidade gestora"].fillna(value= 0000)

In [199]:
df_completa["código órgão superior"].isna().sum()

np.int64(0)

In [200]:
df_completa["origem receita"].unique()

array(['Outras Receitas Correntes', 'Receita de Serviços',
       'Transferências de Capital', 'Alienação de Bens',
       'Receita Agropecuária', 'Transferências Correntes',
       'Receita Patrimonial', 'Operações de Crédito',
       'Impostos, Taxas e Contribuições de Melhoria', 'Sem informação',
       'Amortizações de Empréstimos', 'Contribuições',
       'Receita Industrial', 'Outras Receitas de Capital',
       'Receitas Correntes - a classificar'], dtype=object)

**Creación de nueva columna de porcentajes**

In [223]:
#df_completa.drop(columns= ["percentual realizado"], inplace= True)

In [217]:
def calcular_porcentaje(realizado, previsto):
    if pd.isna(realizado) or pd.isna(previsto):
        return 0
    try:
        return (realizado / previsto) * 100 if previsto != 0 else 0
    except Exception as e:
        return 0  

df_completa['porcentaje ejecutado'] = df_completa.apply(
    lambda row: calcular_porcentaje(row['valor realizado'], row['valor previsto atualizado']),
    axis=1
)


In [219]:
df_completa.shape #732939

(1026139, 16)

**Proporcion nulos en mergeado:**  

- Tenemos 1026138 filas*16 columnas= 16.418.208 celdas de datos

- De esas celdas 140.825 tienen valores nulos lo que supone un 0,86% lo que no es casi nada proporcionalmente

In [220]:
df_completa.isna().sum()          #Tenemos 457.535 nulos 
                                  #Ahora tenemos 433.580 nulos
                                  #hemos conseguido tener 414.602 nulos
                                  #Tras juntar nan y sem informaçao tenemos 338.322 nulos
                                  #Tras crear valores 0000 y sem informaçao tenemos 164.959 nulos
                                  #Tras crear una columna propia conseguimos tener 140.825 nulos

código órgão superior            0
nome órgão superior              0
código órgão                     0
nome órgão                       0
código unidade gestora           0
nome unidade gestora             0
categoria econômica              0
origem receita                   0
espécie receita                  0
detalhamento                     0
valor previsto atualizado    51315
valor lançado                26419
valor realizado              39419
data lançamento              23672
ano exercício                    0
porcentaje ejecutado             0
dtype: int64

In [222]:
df_completa.sample(3)

código órgão superior     nome órgão superior  código órgão  \
228311                52000.0    Ministério da Defesa       52904.0   
427718                25000.0  Ministério da Economia       25000.0   
499017                    0.0          Sem informação       26262.0   

                                               nome órgão  \
228311                                  Fundo do Exército   
427718  Ministério da Economia - Unidades com vínculo ...   
499017                  Universidade Federal de São Paulo   

        código unidade gestora                       nome unidade gestora  \
228311                167086.0                          FUNDO DO EXERCITO   
427718                170013.0    SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
499017                153031.0  UNIFESP-UNIVERSIDADE FEDERAL DE SAO PAULO   

       categoria econômica                               origem receita  \
228311  Receitas Correntes                    Outras Receitas Correntes   
427718  Receitas Correntes  Impostos, Taxas e Contribuições de Melhoria   
499017  Receitas Correntes                          Receita de Serviços   

                                     espécie receita  \
228311                     Demais receitas correntes   
427718                                         Taxas   
499017  Serviços Administrativos e Comerciais Gerais   

                                       detalhamento  \
228311          OUTRAS RECEITAS-PRIMARIAS-PRINCIPAL   
427718   TX.CONTR.E FISC.PESCA E AQUICULTURA-PRINC.   
499017  SERV.ADMINISTRAT.E COMERCIAIS GERAIS-PRINC.   

        valor previsto atualizado  valor lançado  valor realizado  \
228311                        0.0            0.0          4657.93   
427718                        NaN            0.0         16448.84   
499017                        NaN            0.0         11851.39   

       data lançamento  ano exercício  porcentaje ejecutado  
228311      2017-03-16         2017.0                   0.0  
427718      2018-08-17         2018.0                   0.0  
499017      2018-08-09         2018.0                   0.0

**Tratamiento de NULOS**

Dado que la porporcion de nulos llega al 0,8%, merecería la pena eliminarlos y quedarnos con el otro 99,2% pues la pérdida de información no sería masiva.
Pero debido a que se elimina la fila entera, realmente esto supondría perder el 13% de las filas. Lo que creo que se puede asumir.


Esta asunción, se debe a que la opción de rellenar los nulos en los valores numéricos con la media o mediana, aunque es verdad que no afectaría a la media y la mediana, si que perderíamos tanto la información sobre cómo se distribuyen como también  crearía una distorsión en las sumas totales.

14.276.416 celdas completas de info sin ningun nan 
16.418.208 celdas totales (contando nulos)
Perdemos 13% de las celdas

In [203]:
df_completa.isna().sum()

código órgão superior            0
nome órgão superior              0
código órgão                     0
nome órgão                       0
código unidade gestora           0
nome unidade gestora             0
categoria econômica              0
origem receita                   0
espécie receita                  0
detalhamento                     0
valor previsto atualizado    51315
valor lançado                26419
valor realizado              39419
percentual realizado         24134
data lançamento              23672
ano exercício                    0
dtype: int64

In [226]:
dfcompletasinNan= df_completa.dropna()

In [228]:
dfcompletasinNan.shape  #14.276.416 celdas completas  
                        #16.418.208 celdas totales (contando nulos)
                        #Perdemos 13% de las celdas

(892276, 16)

In [229]:
dfcompletasinNan.isna().sum()

código órgão superior        0
nome órgão superior          0
código órgão                 0
nome órgão                   0
código unidade gestora       0
nome unidade gestora         0
categoria econômica          0
origem receita               0
espécie receita              0
detalhamento                 0
valor previsto atualizado    0
valor lançado                0
valor realizado              0
data lançamento              0
ano exercício                0
porcentaje ejecutado         0
dtype: int64

In [ ]:
df_completa.to_csv("datos/netflix_concat.csv") 
dfcompletasinNan.to_csv("datos/netflix_concat.csv") 

**Intento de la visualización**


Para filtrar los nulos en texto sera añadir na= False (si no los queremos) o na= True (si SI los queremos)